### ISSUE TITLE
[OTBN] Zero flag behavior: with or without carry?

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#3336]
The OTBN spec says about the zero (Z) flag: "Set to 1 if the result of the last operation was zero; otherwise 0."

In the old simulator, "result" means the "returned result without the carry" (i.e. only the lower 256 bit).

In the new simulator, the implementation checks all calculated bits, including the carry, against zero.

Which behavior is correct? I'm assuming the old simulator behavior?

We need to clarify the spec here, and adjust at least one implementation.

**Comment by imphil** [comment]
@felixmiller

**Comment by felixmiller** [comment]
TL;DR: Yes, only checking the 256 result bits would be the right approach. 

Some rationale:
For `add` and `addc`:
If you do consider the (resulting) carry for determining the state of the resulting Z flag there would be only one case when the Z flag can be set and that is when both input operands are 0, since the carry can't overflow again (no 257 bit overflow), since both of the operands are 256 bit only. You could argue that you could want to use this methods, to check whether two numbers are both zero with a single instruction but you would rather use an OR for this

Then you could argue that you could get a false positive for a zero-check, namely if the sum of the numbers you add is exactly 2^256. But when would you expect such a sum to be zero at all, only if you want to:

1.   add a negativ and a positive number
2.   do a modular operation
3.   check for a specific limb being zero in a multi-limb operation

1) We don't have real support for negative numbers (e.g. no sign extend I think) but negative numbers can exist after a `sub` operation or as a constant. But when adding a negative number and a positive one, and the result is positive (including the zero here), the carry will be always set, the zero is only reached via a 256 bit overflow. Therefore, including the carry when determining Z here would generate the wrong result.
 2) Additions mod 256 are a typical thing but also here, if there was an overflow, the carry will be set. You could argue but what if you wanted to do mod 257 (which would not be too weird in crypto code) but again, since the operands are only 256 bit the carry itself can never overflow and you can't do mod 257 with a single add.
3) is actually the same as 2)

For `sub`it is more tricky: Everything is always fine for a normal `sub` and also for `subb`as long as the input carry (borrow flag) is 0. Then you can never reach the case (with any input combinations) where the result carry would be set but the carryless result would be 0. This is different if you do a multi limb sub. E.g. you do c=a-b with a=0 and b=(2^256-1)*2^256+1 (which is a 512 bit number). The result as an unsigned number in limbs would be [1,0,1] (or [0,1] and + carry bit being set). Concluding from the Z flag here that the result of the operation would be 0 and the numbers are identical would be wrong here. You had to look at both flags in this case or you had to treat it as a 3-limb operation (which would be the more straight forward approach since the result is actually out of range for two limbs)
